pkg/controller/remotecluster/rcmgr.go

```diff
@@ -0,0 +1,132 @@
+/*
+  Copyright 2021 The Hybridnet Authors.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+package remotecluster
+
+import (
+	"context"
+	"fmt"
+
+	networkingv1 "github.com/oecp/rama/pkg/apis/networking/v1"
+	"github.com/oecp/rama/pkg/rcmanager"
+	"github.com/oecp/rama/pkg/utils"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/klog"
+)
+
+func (c *Controller) startRemoteClusterMgr(clusterName string) error {
+	klog.Infof("processNextRemoteClusterMgr name=%v", clusterName)
+	rcManager, exists := c.rcMgrCache.Get(clusterName)
+	if !exists {
+		klog.Errorf("Can't find rcManager. clusterName=%v", clusterName)
+		return errors.Errorf("Can't find rcManager. clusterName=%v", clusterName)
+	}
+	rcManager.Run()
+	return nil
+}
+
+// use remove+add instead of update
+func (c *Controller) addOrUpdateRCMgr(rc *networkingv1.RemoteCluster) error {
+	// lock in function range to avoid renewing cluster manager when newing one
+	c.rcMgrCache.RWMutex.Lock()
+	defer c.rcMgrCache.RWMutex.Unlock()
+	klog.Infof("[addOrUpdateRCMgr] cluster=%v", rc.Name)
+
+	clusterName := rc.Name
+	if mgr, exists := c.rcMgrCache.rcMgrMap[clusterName]; exists {
+		klog.Infof("Delete cluster %v from cache", clusterName)
+		mgr.Close()
+		delete(c.rcMgrCache.rcMgrMap, clusterName)
+	}
+
+	rcMgr, err := rcmanager.NewRemoteClusterManager(rc, c.kubeClient, c.ramaClient, c.remoteSubnetLister,
+		c.localClusterSubnetLister, c.remoteVtepLister)
+
+	conditions := make([]networkingv1.ClusterCondition, 0)
+	if err != nil || rcMgr == nil || rcMgr.RamaClient == nil || rcMgr.KubeClient == nil {
+		connErr := errors.Errorf("Can't connect to remote cluster %v", clusterName)
+		c.recorder.Eventf(rc, corev1.EventTypeWarning, "ErrClusterConnectionConfig", connErr.Error())
+		conditions = append(conditions, utils.NewClusterOffline(connErr))
+	} else {
+		conditions = CheckCondition(c, rcMgr.RamaClient, rc.ClusterName, DefaultChecker)
+		rc.Status.UUID = rcMgr.ClusterUUID
+	}
+	rc.Status.Conditions = conditions
+
+	_, err = c.ramaClient.NetworkingV1().RemoteClusters().UpdateStatus(context.TODO(), rc, metav1.UpdateOptions{})
+	if err != nil {
+		runtime.HandleError(err)
+		return err
+	}
+	rcMgr.SetIsReady(IsReady(conditions))
+
+	c.rcMgrCache.rcMgrMap[clusterName] = rcMgr
```

Comment:
Maybe use `rcMgrCache.Set()` instead.

Reply:
I didn't do this because of the lock at first. If use the `Set()` function, the RWmutex will be locked twice and thus cause deadlock